go/src/koding/config/genconfig.go

```diff
@@ -0,0 +1,233 @@
+// +build ignore
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"text/template"
+)
+
+var (
+	input  = flag.String("i", "-", "")
+	output = flag.String("o", "-", "")
+)
+
+var t = template.Must(template.New("").Parse(`// This file is auto-generated. DO NOT EDIT!
```

Comment:
Depending on my knowledge I assume that current implementation of `koding/config` uses `go generate` and `go-bindata` to provide a way of storing build time configuration directly in binary file. This allows us to have all necessary data in a single executable file.

Since the configuration from this package is used by klient we don't have to provide additional config.json files which may become inconsistent with those used by other parts of the infrastructure (like tunnelserver etc.).

So, as you said, we can still use `go generate` + `json.Unmarshal` + `json.Marshal` + `go-bindata`. And yes, this could be even slightly easier than proposed solution. So why did I choose it over JSON-like one:

##### Simplicity:

There is no difference between this PR and JSON solution. If you want to add a new variable, you only need to edit `dontKnowHowToNameIt.coffee` file.

##### Correctness:

Let say we used JSON solution and have statement like this:

```go
endpoint := config.Builtin.Endpoints.URL("tunel", env)
```

we incorrectly wrote `tunel` instead of `tunnel`. This will compile and may even pass the tests.

Auto-generated functions enforce compiler to check this at compile time, so if we made error like above:

```go
endpoint := config.MustTunelURL(env)
```

the code won't even compile unless we have endpoint.tunel variable defined in `dontKnowHowToNameIt.coffee` file.

##### Maintainability:

Another example, when we want to remove an unused variable from existing configuration. In JSON-like solution we just need to grep the code and remove all references to string we want to remove. But, what if we have code like that:

```go
name := blackBox.GetEndpointName() // this can be everything.
endpoint := config.Builtin.Endpoints.URL(name, env)
```

without going deeper in the logic we can't be sure that removing a variable is not going break something. Static-function-like code enforces us to make such logic explicit:

```go
switch blackBox.GetEndpointName() {
	case "tunnel":
		endpoint = config.TunnelURL(env)
	// ...
}
```

so it will be easy to check if our variable is still needed or not.

TL;DR

I used templates and auto-generated files to improve code maintainability and enforce correctness of configuration variables.